src/consolidateListNodes.js

```diff
@@ -0,0 +1,99 @@
+// @flow
+
+import isListNode from './isListNode';
+import {Fragment} from 'prosemirror-model';
+import {Node} from 'prosemirror-model';
+import {Transform} from 'prosemirror-transform';
+
+type JointInfo = {
+  content: Fragment,
+  deleteFrom: number,
+  deleteTo: number,
+  insertAt: number,
+};
+
+// Consolidate list nodes.
+// All adjacent list nodes with the same list type and indent level will be
+// joined into one list node.
+// Note that this transform may change the current user selection.
+export default function consolidateListNodes(tr: Transform): Transform {
+  if (tr.getMeta('dryrun')) {
+    // This transform is potentially expensive to perform, so skip it if
+    // the transform is performed as "dryrun".
+    return tr;
+  }
+
+  // Keep the loop running until there's no more list nodes that can be joined.
+  while (true) {
+    const jointInfo = traverseDocAndFindJointInfo(tr.doc);
```

Comment:
do we need to look at the entire doc?

Reply:
That would be the safest way.
The more efficient way is to start with the very first list found within the text selection but I'd rather do that later because I'd like to work on auto-reordering (of list) first.